src/python/gem5/resources/client_api/atlasclient.py

```diff
@@ -47,22 +77,70 @@ def __init__(self, config: Dict[str, str]):
         self.authUrl = config["authUrl"]
 
     def get_token(self):
-        data = {"key": self.apiKey}
-        data = json.dumps(data).encode("utf-8")
+        return self._atlas_http_json_req(
+            self.url,
+            data_json={"key": self.apiKey},
+            headers={"Content-Type": "application/json"},
+            purpose_of_request="Get Access Token with API key",
+        )["access_token"]
+
+    def _atlas_http_json_req(
+        self,
+        url: str,
+        data_json: Dict[str, Any],
+        headers: dict[str, str],
+        purpose_of_request: Optional[str],
+        max_failed_attempts: int = 4,
+        reattempt_pause_base: int = 2,
+    ) -> Dict[str, Any]:
+        """Sends a JSON object over HTTP to a given Atlas MongoDB server and
+        returns the response. This function will attempt to reconnect to the
+        server if the connection fails a set number of times before raising an
+        exception.
+
+        :param url: The URL to open the connection.
+        :param data_json: The JSON object to send.
+        :param headers: The headers to send with the request.
+        :param purpose_of_request: A string describing the purpose of the
+        request. This is optional. It's used to give context to the user if an
+        exception is raised.
+        :param max_failed_attempts: The maximum number of times to an attempt
+        at making a request should be done before throwing an exception.
+        :param reattempt_pause_base: The base of the exponential backoff -- the
+        time between each attempt.
+
+        **Warning**: This function assumes a JSON response.
+        """
+        data = json.dumps(data_json).encode("utf-8")
 
         req = request.Request(
-            self.authUrl,
+            url,
             data=data,
-            headers={"Content-Type": "application/json"},
+            headers=headers,
         )
-        try:
-            response = request.urlopen(req)
-        except HTTPError as e:
-            self.verify_status_code(e.status)
-            return None
-        result = json.loads(response.read().decode("utf-8"))
-        token = result["access_token"]
-        return token
+
+        for attempt in itertools.count(start=1):
+            try:
+                response = request.urlopen(req)
+                break
+            except Exception as e:
+                if attempt >= max_failed_attempts:
+                    raise AtlasClientHttpJsonRequestError(
+                        client=self,
+                        data=data_json,
+                        headers=headers,
```

Comment:
`header` is being passed as a parameter to `AtlasClientHttpJsonRequestError` but it does not have a parameter named `header`.